plugins/woocommerce-blocks/assets/js/blocks/product-collection/block.json

```diff
@@ -38,14 +38,18 @@
 		"forcePageReload": {
 			"type": "boolean",
 			"default": false
+		},
+		"previewState": {
```

Comment:
Why do we need the attribute? Shouldn't the preview state be an internal state? If it's an attribute, then it's trivial to change it using the `setAttributes` function on a _blockEdit_ filter and this API won't be needed at all. 

Going forward to specific areas where the preview state will be **on by definition**, this will create confusion, especially when copying and pasting the block to other locations. 

cc @samueljseay

Reply:
> Why do we need the attribute? Shouldn't the preview state be an internal state? If it's an attribute, then it's trivial to change it using the setAttributes function on a blockEdit filter and this API won't be needed at all.

If thats the case is setAttributes a fine extension point and we don't need this api? 😁 

Overall though, I agree I don't think it needs exposing as an attribute

Reply:
> Why do we need the attribute? Shouldn't the preview state be an internal state? If it's an attribute, then it's trivial to change it using the `setAttributes` function on a _blockEdit_ filter and this API won't be needed at all.

Good question! My initial plan was to keep the state local and not store it in an attribute, but in [this commit](https://github.com/woocommerce/woocommerce/pull/45703/commits/29b1e159ab9f6bff3559bdc53b612fb216ff95b7#diff-3995c6ce9ba16d6b51d0f238b4454f5ad74641cec7c9c0e63a9afde4b8a0dd1b), I replaced the local state with an attribute. 

The main reason behind doing so was that I wanted inner block "Product Template" to have access to preview state , which I did using "usesContext" in block.json file of Product Template block. This is important because I want to re-fetch products whenever preview state changes, which wasn't happening when I was using local state. 

One alternative solution I tried was using React Context. I stored the preview state into React Context and consumed it into "Product Template" block, but It didn't work. Even if the react context was changing, Product Template wasn't re-fetching the products. AFAIR I spent around 1-2 hours on it but wasn't able to make it work. @samueljseay Do you have any hints on what I might have been doing wrong, assuming that I was providing & consuming React Context the right way? Maybe I can create a new branch of what I tried and we can have a quick call to figure out what I might be doing wrong there? Or we can do async, whichever you prefer 🙂

> Going forward to specific areas where the preview state will be **on by definition**, this will create confusion, especially when copying and pasting the block to other locations.

For now, I don't think it would be problematic because I always initialize the preview state to the initial state whenever the Product Collection renders on the Editor side. In other words, if `registerProductCollection` looks like this:
```tsx
registerProductCollection(
	...,
	preview: {
		setPreviewState: ( {
			setState,
			attributes: currentAttributes,
			location,
		}: SetPreviewStateArgs ) => {
			// setPreviewState has access to the current attributes and location.
			console.log( 'setPreviewState' );
			console.log( currentAttributes, location );

			const timeoutID = setTimeout( () => {
				setState( {
					isPreview: false,
					previewMessage: '',
				} );
			}, 5000 );

			return () => clearTimeout( timeoutID );
		},
		initialPreviewState: {
			isPreview: true,
			previewMessage: 'This is in preview mode',
		},
	}
)
```

We will always use `initialPreviewState` on initial load as you can see [here](https://github.com/woocommerce/woocommerce/blob/7802a6d639c913de15e4a0a13ff8b1c20532ae9d/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx#L94-L96). Then, If `setPreviewState` is present then it will run. Therefore AFAIU, copy-pasting won't create any problems.

Reply:
> If thats the case is setAttributes a fine extension point and we don't need this api? 😁

I think we still need to expose `registerProductCollection` as this is the only function that will be used by 3PDs to register their own collections. This will also allow us to provide only single extensibility point which can accept various arguments including `preview`. This way, 3PDs won't need to use multiple extensibility points. What do you think?

However, I agree that it would be great if we don't need to add an attribute for `preview` mode.

Reply:
@xristos3490 I spent good amount of time yesterday & today, trying to use React context instead of attribute to pass the preview state to Product Template block but I wasn't able to make it work. I also tried to debug it with help of @samueljseay but it looks like a rabbit hole. Therefore, I am going to go with attribute solution for now with "__private" suffix as suggested by @sunyatasattva. 

During cooldown, I will work on it again to figure out why React Context isn't working. I will try to reproduce this issue with simple GB blocks. If this issue will be reproducible in GB, then I will report it on GB repo, otherwise I will try to debug it again. Meanwhile, to unblock this PR I will go with the current solution with "__private" suffix with the attribute. Does that sound good to you? 🙂 (You may check [this comment](https://github.com/woocommerce/woocommerce/pull/46369#discussion_r1592388755) for more info)